### ISSUE TITLE
MatrixSymbol inversion fails.

### ISSUE THREAD

**Issue by jmgc** [sympy/sympy#19162]
The code:

```python
X = sp.Matrix(sp.MatrixSymbol('X', 2, 2))
X.inv()
```

generates the error:

```
TypeError: 'Zero' object is not subscriptable
```

**Comment by vanillafrosty** [comment]
Fixed in PR #19217 

You can do this now:
A = MatrixSymbol('A', 2, 2)
A = A.as_explicit()
A.inv()

**Comment by oscarbenjamin** [comment]
I think that #19217 is not really fixing the problem.

The actual problem is not really to do with inverse but somehow passing matrix elements to `equals`:
```julia
In [30]: M = MatrixSymbol('M', 2, 2)                                                                 

In [31]: d = M.as_explicit().det()                                                                   

In [32]: d                                                                                           
Out[32]: M₀₀⋅M₁₁ - M₀₁⋅M₁₀

In [33]: d.equals(0)                                                                                 
---------------------------------------------------------------------------
TypeError 
...
TypeError: 'Zero' object is not subscriptable
```
Somehow internally `d` ends up having `M` substituted for `1` or `0`.

Oh, that's because `.equals` tries randomly substituting for free symbols:
```julia
In [37]: M[0, 0].free_symbols                                                                        
Out[37]: {M}

In [38]: Ms, = M[0, 0].free_symbols                                                                  

In [39]: M[0, 0].subs(Ms, 1)                                                                         
Out[39]: (1)[0, 0]
```
I guess there should be more checking in the `MatrixElement` constructor:
```julia
In [42]: from sympy.matrices.expressions.matexpr import MatrixElement                                

In [43]: MatrixElement(1, 0, 0)                                                                      
Out[43]: (1)[0, 0]

In [44]: MatrixElement(M, 0, 0)                                                                      
Out[44]: M₀₀
```

**Comment by vanillafrosty** [comment]
I see. Originally my thought was to route the inv() method on ImmutableDenseMatrix instances back to the inverse method of the MatrixExpr class (when faced with inverting an ImmutableDenseMatrix formed from a symbolic matrix). My reasoning is that when you construct a symbolic matrix and immediately take the inverse, the .inv() method goes to that of the MatrixExpr class.

I.e.
A = Matrix('A', 2, 2)
A.inv()

But @oscarbenjamin you're saying it's better to fix MatrixElement instead of simply monkey patching class methods? I would agree. Let me take another look

**Comment by vanillafrosty** [comment]
Looks like `.equals` is calling `.is_constant` which is doing the substitution of the symbols @oscarbenjamin mentioned. I don't think it makes sense to substitute the symbols out in a `MatrixElement` for 0 or 1 and then use the instance for comparisons. So to me the best solution is to catch the TypeError thrown after the substitution when the code tries to call `.equals` again. We should still allow the substitution initially because the method is not only going to be used on symbolic instances. 

See my latest commit. I'm catching the error thrown by line 691 in `sympy/core/expr.py`:
`if b is not None and b is not S.NaN and b.equals(a) is False:`

Luckily, regardless of whether this line throws an exception, the code reassigns `b` and we don't worry about it anymore. Hopefully it's as simple as that.

**Comment by GarthSnyder** [comment]
I think I'm running into this same issue, but I'm not entirely sure. Could you verify?

Just from playing around with various configurations, it seems to be the matrix inversion that poisons the well. Test case:

```
from sympy import *

# Moore-Penrose pseudo-inverse
def pinv(matrix):
    return (matrix.T * matrix) ** -1 * matrix.T

n = 5
X = MatrixSymbol("X", n, 1)
Y = MatrixSymbol("Y", n, 1)
points = BlockMatrix([ones(n, 1), X, Y])

pseudo = pinv(points)
pseudo[2, 2]
```
This yields an "Index out of range: a[1]" error with a long traceback.

`points.as_explicit()` is

![Screen Shot 2020-05-17 at 11 33 19 PM](https://user-images.githubusercontent.com/443181/82181171-dab47f80-9896-11ea-9de0-2cc269512a70.png)

which looks correct. `pseudo` is:

![Screen Shot 2020-05-17 at 11 35 08 PM](https://user-images.githubusercontent.com/443181/82181285-12232c00-9897-11ea-8472-7cc5faa28c4a.png)

and `pseudo.shape` is `(3, 5)` which, again, both look fine. It's just that I can't access any actual cell of the result.

If you leave off the multiplication by matrix.T in the pseudo-inverse, you get a Zero-related error message.

**Comment by czgdp1807** [comment]
The error still persists on `master`. Open for fixes.

**Comment by onavhamid** [comment]
Facing same problem, though from a different 2x2 expression using sympy v1.6.2.  Problem persisted in v1.6.0, 1.5.0, but was not present when using v1.4.0.  The original example show by @jmgc runs fine in v1.4.0
![image](https://user-images.githubusercontent.com/54724626/98058658-dbbc7780-1e0a-11eb-927e-7675b15ad188.png)
Would be great to have this get attention.

**Comment by oscarbenjamin** [comment]
I think that the fix should be something like:
```diff
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index 13c5c4a15e..e82373df7a 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -652,7 +652,7 @@ def check_denominator_zeros(expression):
                 if a is S.NaN:
                     # evaluation may succeed when substitution fails
                     a = expr._random(None, 0, 0, 0, 0)
-            except ZeroDivisionError:
+            except (ZeroDivisionError, TypeError):
                 a = None
             if a is not None and a is not S.NaN:
                 try:
diff --git a/sympy/matrices/expressions/matexpr.py b/sympy/matrices/expressions/matexpr.py
index 51db3c3395..fb35561b12 100644
--- a/sympy/matrices/expressions/matexpr.py
+++ b/sympy/matrices/expressions/matexpr.py
@@ -705,7 +705,10 @@ def __new__(cls, name, n, m):
                 return name[n, m]
         if isinstance(name, str):
             name = Symbol(name)
-        name = _sympify(name)
+        else:
+            name = _sympify(name)
+            if not isinstance(name, MatrixExpr):
+                raise TypeError("First argument of MatrixElement should be a matrix")
         obj = Expr.__new__(cls, name, n, m)
         return obj
```

**Comment by czgdp1807** [comment]
The bug still persists on `master`. Open for work.

**Comment by sidhu1012** [comment]
I'll work on it.

**Comment by ri4a** [comment]
Persists in 1.7.1

**Comment by oscarbenjamin** [comment]
I think that the fix is in #20691 from @sidhu1012. The problem is it causes some other test to fail but I think maybe that other test can just be made XFAIL because it's more important to fix this.

**Comment by sidhu1012** [comment]
I'll make that test XFail
